### ISSUE TITLE
Performance problems at 0178f8632521dd97339fde7c06f90760de6455bb

### ISSUE THREAD

**Issue by squareproton** [veny/GearmaNode#26]
We've just finished replacing node-gearman with GearmaNode in our app but in testing we noticed a cpu load increase and a massive drop in performance.

I've run some benchmarks with a php server and a node-gearman vs GearmaNode client and the results are very interesting.

For small payload sizes node-gearman is competitive but (~25% slower) but when the payload gets to about about 10K things start to going south quickly. By the time the job payload is 1MB GearmaNode is processing about ~3 jobs / sec vs node-gearman at ~140 jobs / sec.

The naive benchmarking scripts I used are below.

### client
```javascript
var numCompleted = 0;
var bytesRecieved = 0;

function completeNodeGearman (data) {
    numCompleted++;
    bytesRecieved+= data.length;
}

function completeGearmaNode () {
    numCompleted++;
    bytesRecieved+= this.response.length;
}

setTimeout(function() {
    function fileSizeSI(a,b,c,d,e){
         return (b=Math,c=b.log,d=1e3,e=c(a)/c(d)|0,a/b.pow(d,e)).toFixed(2)+' '+(e?'kMGTPEZY'[--e]+'B':'Bytes')
    }
    console.log(
        "completed %d jobs and %s of payload",
        numCompleted,
        fileSizeSI(bytesRecieved)
    );
    process.exit();
}, 10*1000);

var jobName;
jobName = 'cnt';
jobName = '1K';
jobName = '10K';
jobName = '100K';
jobName = '1M';

if (false) {
    var gearmanode = require('gearmanode');
    gearmanode.Client.logger.transports.console.level = 'error';
    var client = gearmanode.client();
    setInterval(function() {
        var job = client.submitJob(jobName, "");
        job.on('complete', completeGearmaNode);
    }, 1);
} else {
    var NodeGearman = require("node-gearman");
    var nodegearman = new NodeGearman();
    setInterval(function() {
        var job = nodegearman.submitJob(jobName, "");
        job.on("data", completeNodeGearman);
    }, 1);
}

// cnt - GearmanNode ~340 jobs / second, node-gearman ~450 jobs / second
// 1K - GearmanNode ~340 jobs / second, node-gearman ~450 jobs / second
// 10K - GearmanNode ~300 jobs / second, node-gearman ~450 jobs / second
// 100K - GearmanNode ~34 jobs / second, node-gearman ~340 jobs / second
// 1M - GearmanNode ~3 jobs / second, node-gearman ~140 jobs / second
```

### server

```php
<?php

$worker= new GearmanWorker();
$worker->addServer();

$c = 0;
$worker->addFunction("cnt", function ($job) use (&$c) {
    $c++;
    print("job {$c}\n");
    return $c;
});

$onek = str_pad('', 1024);
$worker->addFunction("1K", function ($job) use ($onek) {
    print("job 1K\n");
    return $onek;
});

$tenk = str_pad('', 1024*10);
$worker->addFunction("10K", function ($job) use ($tenk) {
    print("job 10K\n");
    return $tenk;
});

$hundredk = str_pad('', 1024*100);
$worker->addFunction("100K", function ($job) use ($hundredk) {
    print("job 100K\n");
    return $hundredk;
});

$onem = str_pad('', 1024*1024);
$worker->addFunction("1M", function ($job) use ($onem) {
    print("job 1M\n");
    return $onem;
});

while ($worker->work()) {
    continue;
}
```

Our app spends most of it's time processing around 30 jobs per second but this sometimes bursts to 500+. Payloads are typically small or null but infrequently we push payloads of ~0.1-5MB which is absolutely trashing performance.

**Comment by veny** [comment]
Performance optimization can be a time consuming task. I focus on it during weekend.

**Comment by squareproton** [comment]
Thanks Veny - really appreciated.

**Comment by squareproton** [comment]
I think 90% of this is related to the very extensive logging. If you all the calls to winston .log() are removed performance improves massively. In the worst case it appears GearmaNode is only 25% slower than nodegearman when you do this.

**Comment by veny** [comment]
yes, you are right, I discovered it too.
I just released v0.2.2 with corresponding fix.
